### PR TITLE
Add KF functionality

### PR DESCRIPTION
**Pull request by jimboca**
Hey Glen, let me know what you think.  I've tried to follow your coding methods and style.  I'll take any comments or corrections you would like.
One thing I wasn't sure about is the keypad sends the kf but which returns chime mode for the areas, so keypads sends a kf for keypad zero which returns the KF for all areas.
Thanks,
Jim

### CODE REVIEW

**Comment by gwww** [comment]
First, thanks! This is a great addition!

I just approved running the workflow, which will come up with some stuff to clean up.

Once that's done I will take a closer look. A couple of first observations though. I lean to less comments than more and prefer to let the reader refer to the original source... the ElkM1 ASCII reference document. The other thing is break out any code changes not related to the KF changes. The one that struck me was the change of the exception. Last, and I get if this doesn't make sense, are there any tests that can be added?

WRT your last email about setting the desired chime mode by looping through until the right one is hit, I think that's better outside of this library. This library tries to be more about reliable protocol level access to the ElkM1 and not application specific access.

File: elkm1_lib/notify.py
Comment (gwww) [inline elkm1_lib/notify.py:41]: Break this out into a separate PR. That will allow for focused testing. Perhaps add a test (I notice I missed added a test for this).

Comment (jimboca) [reply elkm1_lib/notify.py:41]: Ok, will do.  Yes, that exception drove me crazy for a long time since I assumed it was something I broke in my library, but was really caused by parameter calling changing for the kc, as soon as I added more info to the exception it was obvious where my issue was.

**Comment by jimboca** [comment]
Thanks!  I will work on cleaning this up tonight.   I understand about the comments, and get your reasoning.  I added it for my sanity while creating the code, so it's not necessary now and I will remove.  I did include more KC parsing in this, along with adding the KF parsing.  In reality I'm not sure I need the extra KC stuff, but may in the future, so I could pull that out for now as well if you prefer.

**Comment by gwww** [comment]
I, mostly successfully, follow the YAGNI principle.  So yes, less code is preferred.

**Comment by jimboca** [comment]
Now only contains KF changes, critique as needed please.

**Comment by jimboca** [comment]
In regards to adding tests, I wasn't sure how to structure it.  If there was a test for last keypress I could probably follow that but I don't see it?

**Comment by gwww** [comment]
I know that I approved you to run workflows but for some reason when I checked in on your pushes today you were not approved. Could be I don't understand how to approve or that it was revoked. Anyway, I approved again and the build is failing. The PR will need to pass the build checks. I'll add a couple of comments to start but will do a more full review when the build passes.

File: elkm1_lib/areas.py
Comment (gwww) [inline elkm1_lib/areas.py:137]: I don't understand the reasoning for this try/except. Seems to be checking if the `mode` is in range, but then nothing is done with the result of the check.

File: elkm1_lib/const.py
Comment (gwww) [inline elkm1_lib/const.py:205]: Can simplify the comment; it really just repeats the code below.

File: elkm1_lib/const.py
Comment (gwww) [inline elkm1_lib/const.py:218]:
```suggestion
    """Chime mode settings on an Area"""
```

File: elkm1_lib/keypads.py
Comment (gwww) [inline elkm1_lib/keypads.py:56]: Yup, send for each keypad.

File: elkm1_lib/keypads.py
Comment (gwww) [inline elkm1_lib/keypads.py:86]: Similar comment as before. Passing `name = ""` probably not right thing to do.

File: elkm1_lib/message.py
Comment (gwww) [inline elkm1_lib/message.py:620]: Move this function in sorted order.

**Review by gwww — CHANGES_REQUESTED** [review]

File: elkm1_lib/areas.py
Comment (jimboca) [reply elkm1_lib/areas.py:137]: Yes, I wasn't to sure why it was done that way, but I duplicated what is in kc_handler https://github.com/gwww/elkm1/blob/main/elkm1_lib/keypads.py#L56

File: elkm1_lib/keypads.py
Comment (jimboca) [reply elkm1_lib/keypads.py]: Committed untested fix, will test tonight.

**Comment by jimboca** [comment]
Thanks, some fixes done, it's still waiting for a new workflow approval, last one was clean.

**Comment by gwww** [comment]
I'm still trying to figure out how to give you workflow approval. First time doing it. I figured pressing "Approve and run" was enough, but its not. I just tried configuring something on the repo to see if that helps.

File: elkm1_lib/areas.py
Comment (jimboca) [reply elkm1_lib/areas.py:137]: So in keypads kc_handler it uses name:
`        self.elements[keypad].setattr("last_keypress", (name, key), True)`
But in areas kf_handler I didn't return name:
`            self.elements[area].setattr("chime_mode", mode)`
since it only cares about chime_mode in area.

Comment (jimboca) [reply elkm1_lib/areas.py:137]: So the reason it's different is in kf_decode I handled converting it to an enum, should I change this to just return the int and leave it up to handler to convert to enum?
```
def kf_decode(msg: str) -> dict[str, Any]:
    """KF: Keypad function key press."""
    return {
        "keypad": int(msg[4:6]) - 1,
        "key": FunctionKeys(msg[6]),
        "chime_mode": [ChimeMode(int(x)) for x in msg[7:15]],
    }
```

Comment (jimboca) [reply elkm1_lib/areas.py:136]: it now passes name thru like your kc_decode.

File: elkm1_lib/keypads.py
Comment (jimboca) [reply elkm1_lib/keypads.py:86]: works like kc_decode, let me know if you want all to change to work differently?

**Comment by jimboca** [comment]
I've converted the kf_handler to look like your kc_handler now, so let me know what you think.

**Comment by gwww** [comment]
I made a tweak to the workflow. Hopefully I don't need to keep approving your workflow run on push. So you know, once its clean I'll do a deeper review. That means black, mypy, isort, lint, and pytest all run.

File: elkm1_lib/areas.py
Comment (gwww) [inline elkm1_lib/areas.py:137]: I see now what this is doing. All good.

File: elkm1_lib/keypads.py
Comment (gwww) [inline elkm1_lib/keypads.py:30]: Should `key` be type `FunctionsKeys`?

File: elkm1_lib/keypads.py
Comment (gwww) [inline elkm1_lib/keypads.py:34]: No. Do you even need this function? Could it be left as `press_function_key(FunctionKeys.CHIME)`?

File: elkm1_lib/message.py
Comment (gwww) [inline elkm1_lib/message.py:495]: This is the only encode function that has a default value. What's the thinking here? 

Change type to `FunctionKeys` and then people are forced to pass in a valid function key.

**Review by gwww — CHANGES_REQUESTED** [review]

File: elkm1_lib/areas.py
Comment (jimboca) [reply elkm1_lib/areas.py:137]: great, thanks

File: elkm1_lib/message.py
Comment (jimboca) [reply elkm1_lib/message.py:495]: I added the default so keypads->sync could just call with no arguments and not need to know that null should be sent:
https://github.com/jimboca/elkm1/blob/master/elkm1_lib/keypads.py#L51
Will change the type.

**Comment by gwww** [comment]
If you want to run black, etc, I think this is close.

https://github.com/gwww/elkm1/runs/7714381681?check_suite_focus=true

**Comment by jimboca** [comment]
> If you want to run black, etc, I think this is close.
> 
> https://github.com/gwww/elkm1/runs/7714381681?check_suite_focus=true

**Comment by jimboca** [comment]
Not sure how I closed this....

File: elkm1_lib/keypads.py
Comment (gwww) [inline elkm1_lib/keypads.py:26]: The values you have are good. The general pattern is `None` unless there's something that will cause more problems.

Comment (jimboca) [reply elkm1_lib/keypads.py:26]: Using None is ok, just requires checking for that when referencing it, since once it's defined we get an array.  I'll remove the comment.

File: elkm1_lib/keypads.py
Comment (gwww) [inline elkm1_lib/keypads.py:81]: I think you need to do something similar to the `_kc_handler` where the first line is setting the last_keypress to None. The notifier works on deltas and with key presses the user can press the same key twice in a row and there would only be a single notification.

**Review by gwww — CHANGES_REQUESTED** [review]

File: elkm1_lib/const.py
Comment (gwww) [inline elkm1_lib/const.py:204]: Is something such as `FORCE_KF_SYNC` a better name here?

File: elkm1_lib/keypads.py
Comment (jimboca) [reply elkm1_lib/keypads.py:82]: I wondered about that, change made.

**Comment by jimboca** [comment]
black fixes done

File: elkm1_lib/keypads.py
Comment (gwww) [inline elkm1_lib/keypads.py:13]: No longer needed?

**Comment by gwww** [comment]
I apologize for not being able to turn on the quality check workflow for you. I don't know why it requires me to approve every run. I know for other repos that I contribute to, my PR run automatically. Maybe this will help you check yourself... The workflow bits that need to pass without errors are here: https://github.com/gwww/elkm1/blob/main/.github/workflows/code-quality.yml#L42-L66. You can see each of the commands, in order, that are run to check the code.

**Comment by jimboca** [comment]
No problem.  I think all fixes from your workflow are complete.  Please approve again.

File: elkm1_lib/keypads.py
Comment (gwww) [inline elkm1_lib/keypads.py:23]: `chime_mode` not used on this class.

File: elkm1_lib/keypads.py
Comment (gwww) [inline elkm1_lib/keypads.py:49]: I have tested the code now and it works. I now think I understand why you asked about looping over the keypads here. It does seem redundant.

Here is what I understand. On the keypads `last_function_key` is set to `None`. This sync is called sending `FORCE_KF_SYNC` to each keypad. The responses then set the `last_function_key` to `FORCE_KF_SYNC`.

Meanwhile over in the areas code it receives 16 KF replies and sets the chime mode for each of the 8 areas 16 times.

If that's right, here is what I suggest. In the keypads code, init `last_function_key` to `FORCE_KF_SYNC`. Remove the kf sync code from keypads. Add kf sync code to keypads; a single message will do.

I'm open for discussion, these were my observations from testing.

**Review by gwww — CHANGES_REQUESTED** [review]

**Comment by gwww** [comment]
FYI - I'm travelling starting Thursday and won't be available for about 2 weeks. This PR is real close and we should be able to wrap it today. I will push a new version to PyPI once this is merged.

File: elkm1_lib/keypads.py
Comment (gwww) [inline elkm1_lib/keypads.py:80]: Should this be `(name, key.value)`?

**Comment by jimboca** [comment]
Thanks, I'll try to look at your suggestions today, but it's painful to test the code while at work due to restrictions.  So I may not get a chance until tonight my PDT.

File: elkm1_lib/keypads.py
Comment (jimboca) [reply elkm1_lib/keypads.py:49]: That sounds good

Comment (jimboca) [reply elkm1_lib/keypads.py:49]: Issue moving the send of kf_encode to area, I need the keypad index of one keypads for that area.  I'm not seeing how I can get that from inside he Areas class?

**Comment by jimboca** [comment]
Suggestions done, but untested, with the one caveat mentioned.

File: elkm1_lib/areas.py
Comment (gwww) [inline elkm1_lib/areas.py:96]: I hear you. Maybe just leave a single kf sync in the keypad sync function.

File: elkm1_lib/keypads.py
Comment (jimboca) [reply elkm1_lib/keypads.py:78]: Sure, kc_handler passes the key thru, but it takes in an int for the key. kf_handler takes in an enum and passes that thru.  So it's your your call if it passes thru the enum or the value.

File: elkm1_lib/keypads.py
Comment (gwww) [inline elkm1_lib/keypads.py:73]: Since we only care about "real" function key presses, setting this to `FunctionKeys.FORCE_KF_SYNC` is appropriate. This should address the typing error.

**Review by gwww — CHANGES_REQUESTED** [review]
I just tested and it works as expected. Get a clean build and we're ready to merge. Thanks!

File: elkm1_lib/keypads.py
Comment (jimboca) [reply elkm1_lib/keypads.py]: done

**Comment by jimboca** [comment]
> I just tested and it works as expected. Get a clean build and we're ready to merge. Thanks!

ok, I think it's ready.

**Comment by gwww** [comment]
Thanks!

**Comment by jimboca** [comment]
Thank you!  I appreciate the guidance.

**Comment by gwww** [comment]
Version 2.1.0 published on PyPI.

**Pull request merged**
